client-js/app/scripts/app.ts

```diff
@@ -68,156 +103,378 @@ window.addEventListener('WebComponentsReady', () => {
-      content.removeChild(content.firstChild);
+    p.then(
+      (success) => {
+        newActivity.start().then(
+          (state) => {
+            app.route = state.route;
+            if (state.breadcrumbs) {
+              app.hasBreadcrumbs = true;
+              app.breadcrumbs = state.breadcrumbs;
+            } else {
+              app.hasBreadcrumbs = false;
+            }
+            var content = app.$.content;
+            while (content.firstChild) {
+              content.removeChild(content.firstChild);
+            }
+            content.appendChild(state.elem);
+            activity = newActivity;
+          },
+          (error) => {
+            console.log('error while starting activity', error);
+          }
+        )
+      },
+      (error) => {
+        console.log('error while shutting down activity', error);
+      }
+    )
+  }
+
+  class RegistryActivity implements Activity {
+    path: Array<string>;
+
+    constructor(path: Array<string>) {
+      this.path = path;
+    }
+
+    start(): Promise<ActivityState> {
+      console.log('loading registry:', this.path);
+      return reg.dir({path: this.path}).then((listing) => {
+        var breadcrumbs = [];
+        for (var i = 0; i <= this.path.length; i++) {
+          breadcrumbs.push({
+            name: (i == 0) ? 'registry' : this.path[i-1],
+            isLink: i < this.path.length,
+            url: registryUrl(this.path.slice(0, i))
+          });
+        }
+
+        var dirs = [];
+        for (var i = 0; i < listing.dirs.length; i++) {
+          var dir = listing.dirs[i];
+          dirs.push({
+            name: dir,
+            url: registryUrl(this.path, dir)
+          });
+        }
+        var keys = [];
+        for (var i = 0; i < listing.keys.length; i++) {
+          keys.push({
+            name: listing.keys[i],
+            value: listing.vals[i]
+          });
+        }
+        var elem = <LabradRegistry> LabradRegistry.create();
+        elem.path = this.path;
+        elem.dirs = dirs;
+        elem.keys = keys;
+        elem.socket = reg;
+
+        return {
+          elem: elem,
+          route: 'registry',
+          breadcrumbs: breadcrumbs
+        };
+      });
+    }
+
+    stop(): Promise<void> {
+      return Promise.resolve(null);
     }
-    content.appendChild(elem);
   }
 
-  function pathStr(path: Array<string>, dir?: string): string {
-    var url = '';
-    path.forEach(function(seg) {
-      url += encodeURIComponent(seg) + '/';
-    });
-    if (typeof dir !== 'undefined') {
-      url += encodeURIComponent(dir) + '/';
+  class DatavaultActivity implements Activity {
+    path: Array<string>;
+
+    private elem: LabradGrapher;
+
+    private api: datavault.DataVaultService;
+    private lifetime = new Lifetime();
+
+    constructor(api: datavault.DataVaultService, path: Array<string>) {
+      this.api = api;
+      this.path = path;
+    }
+
+    start(): Promise<ActivityState> {
```

Comment:
This is important and it'd be nice to have a small amount of documentation since this is how we add new interaction with the server. Just something little as a reminder.